src/trio/_core/_asyncgens.py

```diff
@@ -78,12 +80,14 @@ def finalize_in_trio_context(
                 self.trailing_needs_finalize.add(agen)
 
         def finalizer(agen: AsyncGeneratorType[object, NoReturn]) -> None:
-            agen_name = name_asyncgen(agen)
             try:
-                is_ours = not agen.ag_frame.f_locals.get("@trio_foreign_asyncgen")
-            except AttributeError:  # pragma: no cover
+                self.foreign.remove(id(agen))
+            except KeyError:
                 is_ours = True
+            else:
+                is_ours = False
 
+            agen_name = name_asyncgen(agen)
```

Comment:
I moved this call later because it's a LOAD_GLOBAL and fairly complicated bit of code and it's critical that self.foreign.remove get called